### PR TITLE
[RTM] Added basic authentication to app_dev.php

### PR DESCRIPTION
**Pull request by aschempp**
As discussed, see https://github.com/contao/symfony-todo/issues/3

### CODE REVIEW

**Comment by aschempp** [comment]
@discordier is still not watching the repo ;-)

**Comment by leofeyer** [comment]
Haven't we agreed on storing the credentials in a separate file?

**Comment by aschempp** [comment]
No we discussed that but in the end agreed to do the most simple version.

**Comment by leofeyer** [comment]
I recall it differently. @discordier @Toflar What about you?

**Comment by Toflar** [comment]
I don't recall this discussion at all but there's no reason to store it in a different file. Anyway, there's a problem with this PR. See https://github.com/contao/standard-edition/pull/8/files#diff-5f5343d10fae8e62eba1c31ba15d9dc5R38. This will cause my non-Contao API-Requests using Basic Auth to disappear into nirvana.

**Comment by aschempp** [comment]
> This will cause my non-Contao API-Requests using Basic Auth to disappear into nirvana.

Only if you would use the `app_dev.php`. And using Basic Auth would not work anyway, because you would not be given access in the first place… If that would really be the case, you would need to change the `app_dev.php` (same as you would with the standard Symfony version).

**Comment by leofeyer** [comment]
> but there's no reason to store it in a different file

Yes, there is. The file will be marked as modified by Git if I adjust it. We should require people to add a configuration file in `app/config/` instead.

**Comment by Toflar** [comment]
> Yes, there is. The file will be marked as modified by Git if I adjust it. We should require people to add a configuration file in app/config/ instead.

That's a good idea though.

**Comment by aschempp** [comment]
> The file will be marked as modified by Git if I adjust it. 

That's the case with all files of the distribution (e.g. `app/config.yml`). I'm 100% sure we agreed on the most simple solution, I don't know why we discuss that again…

**Comment by leofeyer** [comment]
As discussed on April 27th, we should use SHA512 hashes. We can then add a JavaScript to the user manual, so people can generate the strings and copy and paste them.

**Comment by aschempp** [comment]
The documentation should use http://caligatio.github.io/jsSHA/ to generate the access key from two input fields.

File: web/app_dev.php [hunk 1]
```diff
@@ -11,16 +11,36 @@
...
+        || !isset($_SERVER['PHP_AUTH_PW'])
+        || hash('sha512', $_SERVER['PHP_AUTH_USER']) !== $username
+        || hash('sha512', $_SERVER['PHP_AUTH_PW']) !== $password
+    ) {
+        header('WWW-Authenticate: Basic realm="Contao Debug"');
+        header('HTTP/1.0 401 Unauthorized');
+        exit('You are not allowed to access this file. Check ' . basename(__FILE__) . ' for more information.');
+    }
+
+    unset($username, $password, $_SERVER['PHP_AUTH_USER'], $_SERVER['PHP_AUTH_PW']);
```
Comment (leofeyer) [inline web/app_dev.php]: Why do you unset the global `$_SERVER` variables?

Comment (aschempp) [inline web/app_dev.php]: Because it would contain the plain text password. I think noone should ever use that…

Comment (leofeyer) [inline web/app_dev.php]: But another script may rely on the values being present if a user is authenticated. We should not try to "fix" Apache in our application.

Comment (aschempp) [inline web/app_dev.php]: That's not apache. We'll only remove the values if we successfully authenticated a user using these values.

Comment (leofeyer) [inline web/app_dev.php]: Well, if we don't set these values, we should not remove them, either. @contao/developers What do you think?

Comment (discordier) [inline web/app_dev.php]: I do not like the approach of basic auth in general as digest would be more secure but I am with @leofeyer on the case that we MUST NOT tamper with the contents of `$_SERVER`.

I also am unsure if .htaccess hacks are still needed in cgi environments for PHP.

Further reading: http://php.net/manual/en/features.http-auth.php (example implementation of digest auth included). Also read the comments for some cgi environment hacks.

Comment (aschempp) [inline web/app_dev.php]: I have removed the `$_SERVER` changes.

On the other topics, I know about digest etc., but that's all very much over-engineered for this task. We wanted to add the most simple solution for protected access to the dev controller. Adding anything else would be totally over-engineered for the entry point script. We would then need to add classes and Symfony listeners instead, but I don't think that's a good idea.

**Pull request merged**